# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsContentUtils.cpp
# Commit: 4f30d7afdf22
# Full Hash: 4f30d7afdf22c652481513564b60b37053c0948b
# Author: Atila Butkovits <abutkovits@mozilla.com>
# Date: 2025-05-08 04:32:48
# Regressor Bug: 1932150
# File Overlap Count: 2
# Description:
#   Revert "Bug 1932150 - Fix build bustage" for causing build bustages.
#   
#   This reverts commit 0681b181d6a8a48ac460a97dacb7fca91f9989c3.
#   
#   Revert "Bug 1932150 - Fix lint failure" for causing Build bustage.
# ==============================================================================

diff -r 1e64a48e3517 -r 4f30d7afdf22 dom/base/nsContentUtils.cpp
--- a/dom/base/nsContentUtils.cpp	Wed May 07 20:11:46 2025 +0300
+++ b/dom/base/nsContentUtils.cpp	Wed May 07 20:37:37 2025 +0300
@@ -486,61 +486,31 @@
 int32_t nsContentUtils::sInnerOrOuterWindowCount = 0;
 uint32_t nsContentUtils::sInnerOrOuterWindowSerialCounter = 0;
 
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const RangeBoundary& aFirstBoundary, const RangeBoundary& aSecondBoundary,
     NodeIndexCache* aIndexCache);
-template Maybe<int32_t> nsContentUtils::ComparePoints<TreeKind::Flat>(
-    const RangeBoundary& aFirstBoundary, const RangeBoundary& aSecondBoundary,
-    NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const RangeBoundary& aFirstBoundary,
     const RawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-template Maybe<int32_t> nsContentUtils::ComparePoints<TreeKind::Flat>(
-    const RangeBoundary& aFirstBoundary,
-    const RawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const RawRangeBoundary& aFirstBoundary,
     const RangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-template Maybe<int32_t> nsContentUtils::ComparePoints<TreeKind::Flat>(
-    const RawRangeBoundary& aFirstBoundary,
-    const RangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const RawRangeBoundary& aFirstBoundary,
     const RawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-template Maybe<int32_t> nsContentUtils::ComparePoints<TreeKind::Flat>(
-    const RawRangeBoundary& aFirstBoundary,
-    const RawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
-    const RangeBoundary& aFirstBoundary,
-    const ConstRawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-template Maybe<int32_t> nsContentUtils::ComparePoints<TreeKind::Flat>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const RangeBoundary& aFirstBoundary,
     const ConstRawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const ConstRawRangeBoundary& aFirstBoundary,
     const RangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
+    const RawRangeBoundary& aFirstBoundary,
+    const ConstRawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const ConstRawRangeBoundary& aFirstBoundary,
     const RawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-
-template Maybe<int32_t>
-nsContentUtils::ComparePoints<TreeKind::ShadowIncludingDOM>(
-    const ConstRawRangeBoundary& aFirstBoundary,
-    const ConstRawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
-template Maybe<int32_t> nsContentUtils::ComparePoints<TreeKind::Flat>(
+template Maybe<int32_t> nsContentUtils::ComparePoints(
     const ConstRawRangeBoundary& aFirstBoundary,
     const ConstRawRangeBoundary& aSecondBoundary, NodeIndexCache* aIndexCache);
 
@@ -786,10 +756,9 @@
   return aContent->GetFlattenedTreeParent();
 }
 
-static nsIContent* GetFlattenedTreeParentNodeForSelection(
-    const nsIContent* aNode) {
-  nsINode* parent = aNode->GetFlattenedTreeParentNodeForSelection();
-  return parent && parent->IsContent() ? parent->AsContent() : nullptr;
+static auto* GetFlattenedTreeParentNodeForSelection(
+    const nsIContent* aContent) {
+  return aContent->GetFlattenedTreeParentNodeForSelection();
 }
 
 static auto* GetFlattenedTreeParentElementForStyle(const Element* aElement) {
@@ -802,17 +771,6 @@
              : nullptr;
 }
 
-static bool AreNodesInSameSlot(const nsINode* aNode1, const nsINode* aNode2) {
-  if (auto* content1 = nsIContent::FromNodeOrNull(aNode1)) {
-    if (auto* slot = content1->GetAssignedSlot()) {
-      if (auto* content2 = nsIContent::FromNodeOrNull(aNode2)) {
-        return slot == content2->GetAssignedSlot();
-      }
-    }
-  }
-  return false;
-}
-
 template <typename Node1, typename Node2, typename GetParentFunc>
 class MOZ_STACK_CLASS CommonAncestors final {
  public:
@@ -867,12 +825,9 @@
     return GetClosestCommonAncestorChild(mInclusiveAncestors2);
   }
 
-  template <TreeKind aKind>
   void WarnIfClosestCommonAncestorChildrenAreNotInChildList() const {
-    WarnIfClosestCommonAncestorChildIsNotInChildList<aKind>(
-        mInclusiveAncestors1);
-    WarnIfClosestCommonAncestorChildIsNotInChildList<aKind>(
-        mInclusiveAncestors2);
+    WarnIfClosestCommonAncestorChildIsNotInChildList(mInclusiveAncestors1);
+    WarnIfClosestCommonAncestorChildIsNotInChildList(mInclusiveAncestors2);
   }
 
  private:
@@ -914,9 +869,7 @@
     return child;
   }
 
-  template <TreeKind aKind, typename Node,
-            typename = std::enable_if_t<aKind == TreeKind::ShadowIncludingDOM ||
-                                        aKind == TreeKind::Flat>>
+  template <typename Node>
   void WarnIfClosestCommonAncestorChildIsNotInChildList(
       const nsTArray<Node*>& aInclusiveAncestors) const {
 #ifdef DEBUG
@@ -925,24 +878,8 @@
       if (!child) {
         return;
       }
-
-      if (mClosestCommonAncestor->GetShadowRoot() == child) {
-        return;
-      }
-
-      Maybe<uint32_t> childIndex;
-      if constexpr (aKind == TreeKind::Flat) {
-        if (auto* slot = HTMLSlotElement::FromNode(mClosestCommonAncestor)) {
-          auto index = slot->AssignedNodes().IndexOf(child);
-          if (index != nsTArray<RefPtr<nsINode>>::NoIndex) {
-            childIndex = Some(index);
-          }
-        }
-      }
-
-      if (childIndex.isNothing()) {
-        childIndex = mClosestCommonAncestor->ComputeIndexOf(child);
-      }
+      const Maybe<uint32_t> childIndex =
+          mClosestCommonAncestor->ComputeIndexOf(child);
       if (MOZ_LIKELY(childIndex.isSome())) {
         return;
       }
@@ -3226,7 +3163,6 @@
 }
 
 /* static */
-template <TreeKind aKind, typename Dummy>
 Maybe<int32_t> nsContentUtils::CompareChildNodes(
     const nsINode* aChild1, const nsINode* aChild2,
     NodeIndexCache* aIndexCache /* = nullptr */) {
@@ -3249,23 +3185,6 @@
     MOZ_ASSERT(aChild1->GetParentOrShadowHostNode());
     return Some(-1);
   }
-
-  if constexpr (aKind == TreeKind::Flat) {
-    if (AreNodesInSameSlot(aChild1, aChild2)) {
-      // They differ at slot, so use their position in slot
-      const auto* slot = aChild1->AsContent()->GetAssignedSlot();
-      MOZ_ASSERT(slot);
-
-      auto child1Index = slot->AssignedNodes().IndexOf(aChild1);
-      auto child2Index = slot->AssignedNodes().IndexOf(aChild2);
-
-      MOZ_ASSERT(child1Index != nsTArray<RefPtr<nsINode>>::NoIndex);
-      MOZ_ASSERT(child2Index != nsTArray<RefPtr<nsINode>>::NoIndex);
-
-      return Some(child1Index < child2Index ? -1 : 1);
-    }
-  }
-
   MOZ_ASSERT(aChild1->GetParentOrShadowHostNode());
   const nsINode& commonParentNode = *aChild1->GetParentOrShadowHostNode();
   MOZ_ASSERT(aChild2->GetParentOrShadowHostNode() == &commonParentNode);
@@ -3334,10 +3253,9 @@
 }
 
 /* static */
-template <TreeKind aKind, typename Dummy>
 Maybe<int32_t> nsContentUtils::CompareClosestCommonAncestorChildren(
     const nsINode& aParent, const nsINode* aChild1, const nsINode* aChild2,
-    nsContentUtils::NodeIndexCache* aIndexCache) {
+    nsContentUtils::NodeIndexCache* aIndexCache = nullptr) {
   MOZ_ASSERT_IF(aChild1, GetParentOrShadowHostNode(aChild1));
   MOZ_ASSERT_IF(aChild2, GetParentOrShadowHostNode(aChild2));
 
@@ -3362,7 +3280,7 @@
     return Some(1);
   }
   const Maybe<int32_t> comp =
-      nsContentUtils::CompareChildNodes<aKind>(aChild1, aChild2, aIndexCache);
+      nsContentUtils::CompareChildNodes(aChild1, aChild2, aIndexCache);
   if (MOZ_UNLIKELY(comp.isNothing())) {
     NS_ASSERTION(comp.isSome(),
                  "nsContentUtils::CompareChildNodes() must return Some here. "
@@ -3372,16 +3290,14 @@
     return Some(1);
   }
   MOZ_ASSERT_IF(!*comp, aChild1 == aChild2);
-  MOZ_ASSERT_IF(*comp < 0 && !AreNodesInSameSlot(aChild1, aChild2),
-                (aChild1 ? *aChild1->ComputeIndexInParentNode()
-                         : aParent.GetChildCount()) <
-                    (aChild2 ? *aChild2->ComputeIndexInParentNode()
-                             : aParent.GetChildCount()));
-  MOZ_ASSERT_IF(*comp > 0 && !AreNodesInSameSlot(aChild1, aChild2),
-                (aChild2 ? *aChild2->ComputeIndexInParentNode()
-                         : aParent.GetChildCount()) <
-                    (aChild1 ? *aChild1->ComputeIndexInParentNode()
-                             : aParent.GetChildCount()));
+  MOZ_ASSERT_IF(*comp < 0, (aChild1 ? *aChild1->ComputeIndexInParentNode()
+                                    : aParent.GetChildCount()) <
+                               (aChild2 ? *aChild2->ComputeIndexInParentNode()
+                                        : aParent.GetChildCount()));
+  MOZ_ASSERT_IF(*comp > 0, (aChild2 ? *aChild2->ComputeIndexInParentNode()
+                                    : aParent.GetChildCount()) <
+                               (aChild1 ? *aChild1->ComputeIndexInParentNode()
+                                        : aParent.GetChildCount()));
   return comp;
 }
 
@@ -3433,7 +3349,6 @@
 }
 
 /* static */
-template <TreeKind aKind, typename Dummy>
 Maybe<int32_t> nsContentUtils::ComparePointsWithIndices(
     const nsINode* aParent1, uint32_t aOffset1, const nsINode* aParent2,
     uint32_t aOffset2, NodeIndexCache* aIndexCache) {
@@ -3444,17 +3359,8 @@
     return Some(aOffset1 < aOffset2 ? -1 : (aOffset1 > aOffset2 ? 1 : 0));
   }
 
-  auto GetParentFunc = [](const nsINode* aNode) -> nsINode* {
-    MOZ_ASSERT(aNode);
-    if constexpr (aKind == TreeKind::Flat) {
-      if (aNode->IsContent() && aNode->AsContent()->GetAssignedSlot()) {
-        return aNode->GetFlattenedTreeParentNodeForSelection();
-      }
-    }
-    return aNode->GetParentOrShadowHostNode();
-  };
-
-  const CommonAncestors commonAncestors(*aParent1, *aParent2, GetParentFunc);
+  const CommonAncestors commonAncestors(*aParent1, *aParent2,
+                                        GetParentOrShadowHostNode);
 
   if (MOZ_UNLIKELY(!commonAncestors.GetClosestCommonAncestor())) {
     return Nothing();
@@ -3465,22 +3371,16 @@
   const nsINode* closestCommonAncestorChild2 =
       commonAncestors.GetClosestCommonAncestorChild2();
   MOZ_ASSERT(closestCommonAncestorChild1 != closestCommonAncestorChild2);
-  commonAncestors
-      .template WarnIfClosestCommonAncestorChildrenAreNotInChildList<aKind>();
+  commonAncestors.WarnIfClosestCommonAncestorChildrenAreNotInChildList();
   if (closestCommonAncestorChild1 && closestCommonAncestorChild2) {
-    return CompareClosestCommonAncestorChildren<aKind>(
+    return CompareClosestCommonAncestorChildren(
         *commonAncestors.GetClosestCommonAncestor(),
         closestCommonAncestorChild1, closestCommonAncestorChild2, aIndexCache);
   }
 
   if (closestCommonAncestorChild2) {
-    MOZ_ASSERT(GetParentFunc(closestCommonAncestorChild2) == aParent1);
-    if (aParent1->GetShadowRoot() == closestCommonAncestorChild2) {
-      // Comparing a shadow host with its shadow root.
-      // We consider: [host, 0] < anything in shadow root < [host, 1]
-      return aOffset1 > 0 ? Some(-1) : Some(1);
-    }
-
+    MOZ_ASSERT(closestCommonAncestorChild2->GetParentOrShadowHostNode() ==
+               aParent1);
     // FIXME: bug 1946001, bug 1946003 and bug 1946008.
     if (MOZ_UNLIKELY(
             closestCommonAncestorChild2->IsRootOfNativeAnonymousSubtree() ||
@@ -3507,14 +3407,9 @@
     return comp;
   }
 
-  if (aParent2->GetShadowRoot() == closestCommonAncestorChild1) {
-    // Comparing a shadow host with its shadow root.
-    // We consider: [host, 0] < anything in shadow root < [host, 1]
-    return aOffset2 > 0 ? Some(-1) : Some(1);
-  }
-
   MOZ_ASSERT(closestCommonAncestorChild1);
-  MOZ_ASSERT(GetParentFunc(closestCommonAncestorChild1) == aParent2);
+  MOZ_ASSERT(closestCommonAncestorChild1->GetParentOrShadowHostNode() ==
+             aParent2);
   // FIXME: bug 1946001, bug 1946003 and bug 1946008.
   if (MOZ_UNLIKELY(
           closestCommonAncestorChild1->IsRootOfNativeAnonymousSubtree() ||
@@ -3604,8 +3499,7 @@
 }
 
 /* static */
-template <TreeKind aKind, typename PT1, typename RT1, typename PT2,
-          typename RT2, typename Dummy>
+template <typename PT1, typename RT1, typename PT2, typename RT2>
 Maybe<int32_t> nsContentUtils::ComparePoints(
     const RangeBoundaryBase<PT1, RT1>& aBoundary1,
     const RangeBoundaryBase<PT2, RT2>& aBoundary2,
@@ -3622,7 +3516,7 @@
   // offset in the container.  If both instances have computed offset, we can
   // use ComparePointsWithIndices() which works with offsets.
   if (aBoundary1.HasOffset() && aBoundary2.HasOffset()) {
-    return ComparePointsWithIndices<aKind>(
+    return ComparePointsWithIndices(
         aBoundary1.GetContainer(), *aBoundary1.Offset(kValidOrInvalidOffsets1),
         aBoundary2.GetContainer(), *aBoundary2.Offset(kValidOrInvalidOffsets2),
         aIndexCache);
@@ -3640,27 +3534,18 @@
   if (aBoundary1.GetContainer() == aBoundary2.GetContainer()) {
     const nsIContent* const child1 = aBoundary1.GetChildAtOffset();
     const nsIContent* const child2 = aBoundary2.GetChildAtOffset();
-    return CompareClosestCommonAncestorChildren<aKind>(
-        *aBoundary1.GetContainer(), child1, child2, aIndexCache);
-  }
-
-  auto GetParentFunc = [](const nsINode* aNode) -> nsINode* {
-    MOZ_ASSERT(aNode);
-    if constexpr (aKind == TreeKind::Flat) {
-      if (aNode->IsContent() && aNode->AsContent()->GetAssignedSlot()) {
-        return aNode->GetFlattenedTreeParentNodeForSelection();
-      }
-    }
-    return aNode->GetParentOrShadowHostNode();
-  };
+    return CompareClosestCommonAncestorChildren(*aBoundary1.GetContainer(),
+                                                child1, child2, aIndexCache);
+  }
 
   // Otherwise, we need to compare the common ancestor children which is the
   // most distant different inclusive ancestors of the containers.  So, the
   // following implementation is similar to ComparePointsWithIndices(), but we
   // don't have offset, so, we cannot use offset when we compare the boundaries
   // whose one is a descendant of the other.
-  const CommonAncestors commonAncestors(
-      *aBoundary1.GetContainer(), *aBoundary2.GetContainer(), GetParentFunc);
+  const CommonAncestors commonAncestors(*aBoundary1.GetContainer(),
+                                        *aBoundary2.GetContainer(),
+                                        GetParentOrShadowHostNode);
 
   if (MOZ_UNLIKELY(!commonAncestors.GetClosestCommonAncestor())) {
     return Nothing();
@@ -3671,11 +3556,10 @@
       commonAncestors.GetClosestCommonAncestorChild1();
   const nsINode* closestCommonAncestorChild2 =
       commonAncestors.GetClosestCommonAncestorChild2();
-  commonAncestors
-      .template WarnIfClosestCommonAncestorChildrenAreNotInChildList<aKind>();
+  commonAncestors.WarnIfClosestCommonAncestorChildrenAreNotInChildList();
   MOZ_ASSERT(closestCommonAncestorChild1 != closestCommonAncestorChild2);
   if (closestCommonAncestorChild1 && closestCommonAncestorChild2) {
-    return CompareClosestCommonAncestorChildren<aKind>(
+    return CompareClosestCommonAncestorChildren(
         *commonAncestors.GetClosestCommonAncestor(),
         closestCommonAncestorChild1, closestCommonAncestorChild2, aIndexCache);
   }
@@ -3690,7 +3574,7 @@
       // XXX Keep the odd traditional behavior for now.
       return Some(1);
     }
-    const Maybe<int32_t> comp = nsContentUtils::CompareChildNodes<aKind>(
+    const Maybe<int32_t> comp = nsContentUtils::CompareChildNodes(
         aBoundary1.GetChildAtOffset(), closestCommonAncestorChild2,
         aIndexCache);
     if (NS_WARN_IF(comp.isNothing())) {
@@ -3728,7 +3612,7 @@
     // XXX Keep the odd traditional behavior for now.
     return Some(-1);
   }
-  const Maybe<int32_t> comp = nsContentUtils::CompareChildNodes<aKind>(
+  const Maybe<int32_t> comp = nsContentUtils::CompareChildNodes(
       closestCommonAncestorChild1, aBoundary2.GetChildAtOffset(), aIndexCache);
   if (NS_WARN_IF(comp.isNothing())) {
     NS_ASSERTION(comp.isSome(),